app/src/main/java/com/neeva/app/neevascope/BloomFilterManager.kt

```
@@ -0,0 +1,245 @@
+import java.io.File
+import java.io.InputStream
+import java.net.URL
+import java.nio.charset.Charset
+import java.security.MessageDigest
+import kotlin.experimental.and
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.launch
+import okhttp3.OkHttpClient
+import okhttp3.Request
+
+class BloomFilterManager(
+    private val filesDir: Deferred<File>,
+    private val coroutineScope: CoroutineScope,
+    private val dispatchers: Dispatchers
+) {
+    companion object {
+        private val TAG = BloomFilterManager::class.simpleName
+        private const val DIRECTORY_BLOOM_FILTER = "bloom_filter"
+    }
+
+    @JsonClass(generateAdapter = true)
+    data class CheckSum(
+        @Json(name = "sha") val sha: String,
+        @Json(name = "md5") val md5: String
+    )
+
+    data class BloomFilterConfiguration(
+        val identifier: String,
+        val filterUrl: URL,
+        val checksumUrl: URL,
+        var localUrl: Uri
+    )
+
+    private val reddit = BloomFilterConfiguration(
+        identifier = "reddit",
+        filterUrl = URL("https://s.neeva.co/web/neevascope/v1/reddit.bin"),
+        checksumUrl = URL("https://s.neeva.co/web/neevascope/v1/reddit_latest.json"),
+        // file:///data/user/0/com.neeva.app.debug/cache/bloom_filter/reddit.bin
+        localUrl = Uri.EMPTY
+    )
+
+    val bloomfilter = BloomFilter()
+
+    fun contains(key: String): Boolean? {
+        if (bloomfilter.filter == null) {
+            load()
+            return null
+        }
+
+        return bloomfilter.mayContain(key)
+    }
+
+    private suspend fun getBloomFilterDirectory(): File {
+        return File(filesDir.await(), DIRECTORY_BLOOM_FILTER)
+    }
+
+    suspend fun getFilterBinaryFile(name: String): File {
+        return File(getBloomFilterDirectory(), name)
+    }
+
+    // Download and save the binary file locally
+    fun load() {
+        // check url validity
+        if (
```

`filterUrl` and `checksumUrl` are constants, right?  Should you use `assert` instead?  That way if a developer accidentally changes the URLs to something wrong it'll crash instead of logging a failure.